pkg/auth/providers/ldap/ldap_client.go

```diff
@@ -158,17 +161,108 @@ func (p *ldapProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult
 		//deduplicate groupprincipals get from userMemberAttribute
-		nonDupGroupPrincipals = p.findNonDuplicateGroupPrincipals(newGroupPrincipals, groupPrincipals, nonDupGroupPrincipals)
+		nonDupGroupPrincipals = p.findNonDuplicateBetweenGroupPrincipals(newGroupPrincipals, groupPrincipals, nonDupGroupPrincipals)
 		groupPrincipals = append(groupPrincipals, nonDupGroupPrincipals...)
 		if err != nil {
 			return userPrincipal, groupPrincipals, err
 		}
 	}
+	// Handle nestedgroups for openldap, filter operationalAttrList already handles nestedgroups for freeipa
+	if groupScope == "openldap_group" {
+		searchDomain := config.UserSearchBase
+		if config.GroupSearchBase != "" {
+			searchDomain = config.GroupSearchBase
+		}
+
+		// Handling nestedgroups: tracing from down to top in order to find the parent groups, parent parent groups, and so on...
+		// When traversing up, we note down all the parent groups and add them to groupPrincipals
+		for _, groupPrincipal := range groupPrincipals {
+			err = p.gatherParentGroups(groupPrincipal, searchDomain, groupScope, config, lConn, groupMap, &nestedGroupPrincipals)
+			if err != nil {
+				return userPrincipal, groupPrincipals, nil
+			}
+		}
+		newGroupPrincipals := p.deDupWithInGroupPrincipals(nestedGroupPrincipals)
+		nonDupGroupPrincipals = p.findNonDuplicateBetweenGroupPrincipals(newGroupPrincipals, groupPrincipals, []v3.Principal{})
+		groupPrincipals = append(groupPrincipals, nonDupGroupPrincipals...)
+	}
 
 	return userPrincipal, groupPrincipals, nil
 }
 
-func (p *ldapProvider) findNonDuplicateGroupPrincipals(newGroupPrincipals []v3.Principal, groupPrincipals []v3.Principal, nonDupGroupPrincipals []v3.Principal) []v3.Principal {
+func (p *ldapProvider) gatherParentGroups(groupPrincipal v3.Principal, searchDomain string, groupScope string, config *v3.LdapConfig, lConn *ldapv2.Conn, groupMap map[string]bool, nestedGroupPrincipals *[]v3.Principal) error {
+	groupMap[groupPrincipal.ObjectMeta.Name] = true
+	principals := []v3.Principal{}
+	parts := strings.SplitN(groupPrincipal.ObjectMeta.Name, ":", 2)
+	if len(parts) != 2 {
+		return errors.Errorf("invalid id %v", groupPrincipal.ObjectMeta.Name)
+	}
+	groupDN := strings.TrimPrefix(parts[1], "//")
+
+	searchGroup := ldapv2.NewSearchRequest(searchDomain,
+		ldapv2.ScopeWholeSubtree, ldapv2.NeverDerefAliases, 0, 0, false,
+		fmt.Sprintf("(&(%v=%v)(objectClass=%v))", config.GroupMemberMappingAttribute, ldapv2.EscapeFilter(groupDN), config.GroupObjectClass),
+		p.getGroupSearchAttributes(config), nil)
+	resultGroups, err := lConn.Search(searchGroup)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(resultGroups.Entries); i++ {
+		entry := resultGroups.Entries[i]
+		principal, err := p.attributesToPrincipal(entry.Attributes, entry.DN, groupScope, config)
+		if err != nil {
+			return err
+		}
+		principals = append(principals, *principal)
+		*nestedGroupPrincipals = append(*nestedGroupPrincipals, *principal)
```

Comment:
Can we do the append to nestedGroupPrincipals out of this loop and add it  only when you are not going to skip the group to avoid adding duplicates.